### PR TITLE
[SYCLomatic] Migrate asm 'lop3' ins.

### PR DESCRIPTION
**Pull request by tangjj11**
Signed-off-by: Tang, Jiajun jiajun.tang@intel.com

### CODE REVIEW

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 1]
```diff
@@ -14751,9 +14751,58 @@ void AsmRule::registerMatcher(ast_matchers::MatchFinder &MF) {
       this);
 }
 
+std::string GetAsmLop3Expr(const std::string &TemplateString,
```
Comment (intwanghao) [inline clang/lib/DPCT/ASTTraversal.cpp]: the function name convention is using lower case for first letter

**Comment by tangjj11** [comment]
The introduce of this instruction is here: https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#logic-and-shift-instructions-lop3

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 1 above)
Comment (tangjj11) [reply clang/lib/DPCT/ASTTraversal.cpp]: fixed

**Review by zhiweij1 — APPROVED** [review]

**Review by ziranzha — APPROVED** [review]
LGTM

File: clang/test/dpct/helper_files_ref/include/asm_utils.hpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+#define __DPCT_ASM_UTILS_HPP__
+
+#include <sycl/sycl.hpp>
+
+namespace dpct {
+
+inline uint32_t logical_op_3inputs(uint32_t a, uint32_t b, uint32_t c, uint8_t OpTypes) {
+  uint32_t d = 0;
+  if (OpTypes & 0x01)
+    d |= ~a & ~b & ~c;
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/asm_utils.hpp]: Suggest to remove the if statement,  as the branch code if not effiecent on GPU when execution.

Comment (tangjj11) [reply clang/test/dpct/helper_files_ref/include/asm_utils.hpp]: move this logic out of help function.

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 3]
```diff
@@ -14777,9 +14777,89 @@ void AsmRule::registerMatcher(ast_matchers::MatchFinder &MF) {
...
+
 void AsmRule::runRule(const ast_matchers::MatchFinder::MatchResult &Result) {
-  if (auto E = getNodeAsType<Stmt>(Result, "asm")) {
-    report(E->getBeginLoc(), Diagnostics::DEVICE_ASM, true);
+  if (auto AS = getNodeAsType<AsmStmt>(Result, "asm")) {
+    auto AsmString = AS->generateAsmString(*Result.Context);
+    auto TemplateString = StringRef(AsmString).substr(0, AsmString.find(';'));
+    auto CurrIndex = TemplateString.find(' ');
+    auto OpCode = TemplateString.substr(0, CurrIndex);
+    // TODO: now only support 'lop3' assembly instruction.
```
Comment (zhimingwang36) [inline clang/lib/DPCT/ASTTraversal.cpp]: Please remove the TODO line.

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 4]
```diff
@@ -14777,9 +14777,89 @@ void AsmRule::registerMatcher(ast_matchers::MatchFinder &MF) {
...
+            TemplateString.substr(PreIndex + 1, CurrIndex - PreIndex - 1).str();
+        // Replace all args.
+        if (TempStr.find('$') != TempStr.length() - 2 && Operands.size() != 4) {
+          // When the operands only contain a register, or is the last imm, not
+          // need add the paren.
+          TempStr = "(" + TempStr + ")";
+        }
+        auto ArgIndex = TempStr.find('$');
+        while (ArgIndex != std::string::npos) {
+          // The PTX Instructions has mostly 4 parameters, so just use the char
```
Comment (zhimingwang36) [inline clang/lib/DPCT/ASTTraversal.cpp:14857]: Please add ASM intstruction pattern in commemnt, then easy to read the code.

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 5]
```diff
@@ -14777,9 +14777,89 @@ void AsmRule::registerMatcher(ast_matchers::MatchFinder &MF) {
...
+          Args.push_back("(" + EA.getReplacedString() + ")");
+        }
+      }
+      llvm::SmallVector<std::string, 5> Operands;
+      auto PreIndex = CurrIndex;
+      CurrIndex = TemplateString.find(",", PreIndex);
+      while (PreIndex != StringRef::npos) {
+        auto TempStr =
+            TemplateString.substr(PreIndex + 1, CurrIndex - PreIndex - 1).str();
+        // Replace all args.
```
Comment (zhimingwang36) [inline clang/lib/DPCT/ASTTraversal.cpp]: the comment is not clear, please put ASM pseudo code here to make comment more clear

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 6]
```diff
@@ -669,7 +669,7 @@ void DpctFileInfo::insertHeader(HeaderType Type) {
                           "<" + getCustomMainHelperFileName() +
                               "/ccl_utils.hpp>");
     case HT_Atomic:
-      return insertHeader(HeaderType::HT_CCL, LastIncludeOffset,
+      return insertHeader(HeaderType::HT_Atomic, LastIncludeOffset,
```
Comment (zhimingwang36) [inline clang/lib/DPCT/AnalysisInfo.cpp]: Please summit a seperate PR for this fix.

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 3 above)
Comment (tangjj11) [reply clang/lib/DPCT/ASTTraversal.cpp]: done

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 4 above)
Comment (tangjj11) [reply clang/lib/DPCT/ASTTraversal.cpp:14857]: done

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 5 above)
Comment (tangjj11) [reply clang/lib/DPCT/ASTTraversal.cpp]: done

File: clang/lib/DPCT/AnalysisInfo.cpp (hunk 6 above)
Comment (tangjj11) [reply clang/lib/DPCT/AnalysisInfo.cpp]: done

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**